security/advancedtls/advancedtls.go

```diff
@@ -90,14 +102,24 @@ type ClientOptions struct {
 	// function every time asked to present certificates to the server when a new connection is
 	// established. This is known as peer certificate reloading.
 	GetClientCertificate func(*tls.CertificateRequestInfo) (*tls.Certificate, error)
-	// VerifyPeer is a custom server authorization checking after certificate signature check.
-	// If this is set, we will replace the hostname check with this customized authorization check.
-	// If this is nil, we fall back to typical hostname check.
+	// VerifyPeer is a custom verification check after certificate signature check.
+	// If this is set, we will perform this customized check after doing the normal check(s)
+	// indicated by setting |VType|.
 	VerifyPeer CustomVerificationFunc
 	// ServerNameOverride is for testing only. If set to a non-empty string,
 	// it will override the virtual host name of authority (e.g. :authority header field) in requests.
 	ServerNameOverride string
+	// RootCertificateOptions is REQUIRED to be correctly set on client side.
 	RootCertificateOptions
```

Comment:
Not really related to changed proposed in this PR.
Was there a particular reason to embed this struct  instead of just declaring the two fields directly in `ClientOptions` and `ServerOptions`? The `tls.Config` struct defines these fields separately for the client and server sides. If we remove the embedding, we could name the fields in `ClientOptions` and `ServerOptions` similar to how they are named in `tls.Config`.

Reply:
I don't have strong preferences over the embedding/ non-embedding approach. The current way is actually to distinguish from the way in `tls.Config`. In `tls.Config`, we have `RootCAs`  acts as the client side trust credential, and `ClientCAs` acts as the server side credential, which I thought might be a bit misleading here since the two fields are serving the same purposes, but with different names.
Please let me know if you have any other concerns/suggestions, and I am happy to make another PR for that :-)